x-pack/plugin/core/src/test/java/org/elasticsearch/xpack/core/security/authc/AuthenticationTests.java

```diff
@@ -7,26 +7,150 @@
+        assertFalse(new Authentication(user1, authRealm, realm1).sameUserAs(new Authentication(user1, authRealm, realm1Alike)));
+
+        // File and native realms only needs realm type to match
+        assertTrue(new Authentication(user1, fileRealm1, null).sameUserAs(new Authentication(user1, fileRealm2, null)));
+        assertTrue(new Authentication(user1, authRealm, fileRealm1).sameUserAs(new Authentication(user1, authRealm, fileRealm2)));
+        assertTrue(new Authentication(user1, nativeRealm1, null).sameUserAs(new Authentication(user1, nativeRealm2, null)));
+        assertTrue(new Authentication(user1, authRealm, nativeRealm1).sameUserAs(new Authentication(user1, authRealm, nativeRealm2)));
+
+        // Different usernames
+        assertFalse(new Authentication(user1, realm1, lookupRealm).sameUserAs(new Authentication(user2, realm1, lookupRealm)));
+        // Same username, totally different realm
+        assertFalse(new Authentication(user1, realm1, null).sameUserAs(new Authentication(user1, realm2, null)));
+        // same username, totally different run-as realm
+        assertFalse(new Authentication(user1, randomFrom(realm1, realm2), realm3).sameUserAs(
+            new Authentication(user1, randomFrom(realm1, realm2), realm4)));
+    }
+
+    public void testSameUserAsForApiKeys() {
+        final String username1 = randomAlphaOfLengthBetween(3, 8);
+        final String apiKeyId1 = randomAlphaOfLengthBetween(10, 20);
+        final String apiKeyId2 = randomValueOtherThan(apiKeyId1, () -> randomAlphaOfLengthBetween(10, 20));
+        // Same user, same API key ID
+        assertTrue(randomApiKeyAuthentication(username1, apiKeyId1).sameUserAs(randomApiKeyAuthentication(username1, apiKeyId1)));
+        // same user, different API key ID
+        assertFalse(randomApiKeyAuthentication(username1, apiKeyId1).sameUserAs(randomApiKeyAuthentication(username1, apiKeyId2)));
+
+        // User and its API key
+        final User user2 = randomUserOtherThan(new User[0]);
+        assertFalse(randomApiKeyAuthentication(user2.principal(), apiKeyId1).sameUserAs(
+            new Authentication(user2, randomRealmOtherThan(new RealmRef[0]), null)));
+    }
+
+    public void testSameUserAsForToken() {
+        final User user1 = randomUserOtherThan(new User[0]);
+        final User user2 = randomUserOtherThan(user1);
+        final RealmRef realm1 = randomRealmOtherThan(new RealmRef[0]);
+        final RealmRef realm2 = randomRealmOtherThan(new RealmRef[0]);
+
+        final RealmRef lookupRealm = randomFrom(realm2, null);
+        assertTrue(new Authentication(user1, realm1, lookupRealm).sameUserAs(
+            new Authentication(user1, realm1, lookupRealm, Version.CURRENT, AuthenticationType.TOKEN, Map.of())));
+        assertFalse(new Authentication(user1, realm1, lookupRealm).sameUserAs(
+            new Authentication(user2, realm1, lookupRealm, Version.CURRENT, AuthenticationType.TOKEN, Map.of())));
+    }
+
```

Comment:
the `authnA.sameUser(authnB)` should always return the same result as `authnB.sameUser(authnA)`.
Would be great if you could check for it in tests (at least in the more interesting cases).

Reply:
Good point. It is added as part of the test rewrite. See also above.